src/main/java/net/minecraftforge/fml/common/ILanguageAdapter.java

```diff
@@ -22,17 +22,63 @@
 import java.lang.reflect.Field;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
 
+import net.minecraftforge.fml.common.event.FMLEvent;
 import net.minecraftforge.fml.relauncher.Side;
 
 import org.apache.logging.log4j.Level;
 
-public interface ILanguageAdapter {
+public interface ILanguageAdapter
+{
     public Object getNewInstance(FMLModContainer container, Class<?> objectClass, ClassLoader classLoader, Method factoryMarkedAnnotation) throws Exception;
     public boolean supportsStatics();
     public void setProxy(Field target, Class<?> proxyTarget, Object proxy) throws IllegalArgumentException, IllegalAccessException, NoSuchFieldException, SecurityException;
     public void setInternalProxies(ModContainer mod, Side side, ClassLoader loader);
 
+    /**
+     * Wraps an event handler method in an {@code EventHandler} instance
+     * @param method the eventhandler method of a mod
+     */
+    public default Function<FMLEvent, CompletableFuture<Void>> createEventHandler(Method method)
+    {
+        return event ->
+        {
+            CompletableFuture<Void> future = new CompletableFuture<>();
+            try
+            {
+                if (CompletionStage.class.isAssignableFrom(method.getReturnType()))
+                {
+                    // safely transform the result of the event handler into a CompletableFuture
+                    //noinspection unchecked
+                    ((CompletionStage<Void>) method.invoke(event)).whenComplete((value, throwable) ->
+                    {
+                        if (throwable != null)
+                        {
+                            future.completeExceptionally(throwable);
+                        }
+                        else
+                        {
+                            future.complete(value);
+                        }
+                    });
+                }
+                else
+                {
+                    method.invoke(event);
```

Comment:
The invocation needs a mod instance, I think

Reply:
That is indeed the case, the event surely isn't the instance we are invoking the method on.